src/python/pants/core/goals/test.py

```diff
@@ -171,6 +173,12 @@ def debug_hint(self):
         return self.address.spec
 
 
+@dataclass(frozen=True)
+class TestRequest:
+    field_set: TestFieldSet
+    extra_env_vars: FrozenDict[str, Optional[str]]
```

Comment:
(I recognize you're still iterating)

It'd be great to avoid this extra type imo, if possible, as it adds a new layer of indirection. Instead, could we have a new singleton type `TestExtraEnv`, and rule runners are expected to include that as a param?